src/main/java/core/stack/MateStack.java

```diff
@@ -1,19 +1,63 @@
 package core.stack;
 
+import java.util.EmptyStackException;
+
 public class MateStack<T> {
-    public void push(T value) {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
 
+    public MateStack() {
+        size = 0;
+    }
+
+    public void push(T value) {
+        if (size == 0) {
+            head = new Node<>(value, null);
+            tail = head;
+            size++;
+            return;
+        }
+        tail.next = new Node<>(value, null);
+        tail = tail.next;
+        size++;
     }
 
     public T peek() {
-        return null;
+        checkStack();
+        return tail.value;
     }
 
     public T pop() {
-        return null;
+        checkStack();
+        Node<T> localNode = head;
+        size--;
+        for (int i = 0; i < size - 1; i++) {
+            localNode = localNode.next;
+        }
+        T result = tail.value;
+        tail = localNode;
+        tail.next = null;
+        return result;
```

Comment:
We don't need to iterate since any stack implementation is LIFO data structure, we are always taking the top-most element